xcm/xcm-executor/src/assets.rs

```diff
@@ -216,26 +365,273 @@ impl Assets {
 					// remove the maxmimum possible up to id/amount from self, add the removed onto
 					// result
 					if let Some(entry) = self.non_fungible.take(&(class, instance)) {
-						self.non_fungible.insert(entry);
+						result.non_fungible.insert(entry);
 					}
 				}
-				// TODO: implement partial wildcards.
-				_ => {
-					Default::default()
-				}
-				// MultiAsset::AllFungible
-				// | MultiAsset::AllNonFungible
-				// | MultiAsset::AllAbstractFungible { id }
-				// | MultiAsset::AllAbstractNonFungible { class }
-				// | MultiAsset::AllConcreteFungible { id }
-				// | MultiAsset::AllConcreteNonFungible { class } => (),
 			}
 		}
 		result
 	}
 
+	/// Swaps two mutable Assets, without deinitializing either one.
 	pub fn swapped(&mut self, mut with: Assets) -> Self {
 		swap(&mut *self, &mut with);
 		with
 	}
 }
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	#[allow(non_snake_case)]
+	fn AF(id: u8, amount: u128) -> MultiAsset {
+		MultiAsset::AbstractFungible { id: vec![id], amount }
+	}
+	#[allow(non_snake_case)]
+	fn ANF(class: u8, instance_id: u128) -> MultiAsset {
+		MultiAsset::AbstractNonFungible { class: vec![class], instance: AssetInstance::Index { id: instance_id } }
+	}
+	#[allow(non_snake_case)]
+	fn CF(amount: u128) -> MultiAsset {
+		MultiAsset::ConcreteFungible { id: MultiLocation::Null, amount }
+	}
+	#[allow(non_snake_case)]
+	fn CNF(instance_id: u128) -> MultiAsset {
+		MultiAsset::ConcreteNonFungible { class: MultiLocation::Null, instance: AssetInstance::Index { id: instance_id } }
+	}
+
+	fn test_assets() -> Assets {
+		let mut assets_vec: Vec<MultiAsset> = Vec::new();
+		assets_vec.push(AF(1, 100));
+		assets_vec.push(ANF(2, 200));
+		assets_vec.push(CF(300));
+		assets_vec.push(CNF(400));
+		assets_vec.into()
+	}
+
+	#[test]
+	fn into_assets_iter_works() {
+		let assets = test_assets();
+		let mut iter = assets.into_assets_iter();
+		// Order defined by implementation: CF, AF, CNF, ANF
+		assert_eq!(Some(CF(300)), iter.next());
+		assert_eq!(Some(AF(1, 100)), iter.next());
+		assert_eq!(Some(CNF(400)), iter.next());
+		assert_eq!(Some(ANF(2, 200)), iter.next());
+		assert_eq!(None, iter.next());
```

Comment:
nit: could you `.collect::<Vec<_>>()` and then write it like 

```rust
assert_eq!(assets_vec, vec![CF(300), AF(1, 100), CNF(400), ANF(2, 200)]);
```

looks more legible to me